packages/react-components/react-carousel-preview/library/src/components/useEmblaCarousel.ts

```diff
@@ -0,0 +1,84 @@
+import EmblaCarousel, { type EmblaCarouselType, type EmblaOptionsType } from 'embla-carousel';
+import * as React from 'react';
+
+import { carouselCardClassNames } from './CarouselCard/useCarouselCardStyles.styles';
+import { carouselSliderClassNames } from './CarouselSlider/useCarouselSliderStyles.styles';
+
+const DEFAULT_EMBLA_OPTIONS: EmblaOptionsType = {
+  containScroll: false,
+  inViewThreshold: 0.99,
+  watchDrag: false,
+
+  container: `.${carouselSliderClassNames.root}`,
+  slides: `.${carouselCardClassNames.root}`,
+};
+
+export const EMBLA_VISIBILITY_EVENT = 'embla:visibilitychange';
+
+export function useEmblaCarousel({ align, direction, loop }: Pick<EmblaOptionsType, 'align' | 'direction' | 'loop'>) {
+  const emblaOptions = React.useRef<EmblaOptionsType>({ align, direction, loop });
+  const emblaApi = React.useRef<EmblaCarouselType | null>(null);
+
+  const ref = React.useMemo(() => {
+    let currentElement: HTMLDivElement | null = null;
+
+    const handleVisibilityChange = () => {
+      const cardElements = emblaApi.current?.slideNodes();
+      const visibleIndexes = emblaApi.current?.slidesInView() ?? [];
+
+      cardElements?.forEach((cardElement, index) => {
+        cardElement.dispatchEvent(
+          new CustomEvent(EMBLA_VISIBILITY_EVENT, {
```

Comment:
Please create either a shared type for new event class that extends CustomEvent for this